javascript/organizing_your_javascript_code/factory_functions_and_module_pattern.md - revamps the lesson (#26339)
TheOdinProject/curriculum: javascript/organizing_your_javascript_code/factory_functions_and_module_pattern.md

### Why not constructors?

> We have discussed object constructors in the previous lesson. However, they are simply one of the many ways to organize your code. While they are fairly common in the code seen about the internet and a fundamental building block of the Javascript language, they have their flaws.

Reviewer: ```suggestion
We have discussed object constructors in the previous lesson. However, they are simply one of the many ways to organize your code. While they are fairly common in the code seen about the internet and a fundamental building block of the JavaScript language, they have their flaws.
```

### Lesson overview

This section contains a general overview of topics that you will learn in this lesson.

- Describe the scope of a variable.
- Explore what closures are.
- Briefly consider the disadvantages of using constructors.
- Discuss Factory functions with examples.
- Discuss Private variables and functions concerning factory functions.
- Showcase object inheritance with the help of factory functions.
- Describe what module pattern and IIFEs are.
- Discuss encapsulation and how the module pattern helps with namespacing.

### Scoopfuls of scopes

The word "scoping" essentially asks, "Where is a certain variable available to me?" - it indicates the current context of a variable. When a variable is not declared within **any** functions, existing outside any `{ curly braces }`, they are said to be in the **global scope**, meaning that they are available everywhere. If they are within a function or `{ curly braces }`, they are known to be **locally scoped**.

> Before ECMAScript 6, Javascript had a single keyword to declare a variable, `var`. These variables can be redefined and updated, and are said to be defined within the **function scope**, meaning, they are only available within the function they are declared in.

Reviewer: ```suggestion
Before ECMAScript 6, JavaScript had a single keyword to declare a variable, `var`. These variables can be redefined and updated, and are said to be defined within the **function scope**, meaning, they are only available within the function they are declared in.
```

In ECMAScript 6, the keywords `let` and `const` were introduced. While `var` variables were function scoped, these allow you to define variables that are **block scoped** - basically, scoping the variable to only be available within the closest set of `{ curly braces }` in which it was defined. These braces can be those of a `for` loop, `if-else` condition, or any other similar construct, and are called, a block. Let's see an example to sum this all up.

~~~javascript
let globalAge = 23; // This is a global variable

function printAge (age) { // This is a function - and hey, a curly brace indicating a block
  var varAge = 34; // This is a function scoped variable

  if (age > 0) { // This is yet another curly brace, and thus a block

>     const constAge = age * 2; // This is a block scoped variable that exists in it's closest outer block, the if's in this case

Reviewer: nearest/closest enclosing block might be a better phrase than outer block
```suggestion
    const constAge = age * 2; // This is a block-scoped variable that exists within its nearest enclosing block, which, in this case, is the 'if' block.
```

    console.log(constAge);
  }

>   console.log(constAge); // ERROR! We tried to access a block scoped variable outside its scope

Reviewer: ```suggestion
  console.log(constAge); // ERROR! We tried to access a block scoped variable not within its scope
```

}

printAge(globalAge);

> console.log(varAge); // ERROR! We tried to access a function scoped outside the function it's defined in

Reviewer: ```suggestion
console.log(varAge); // ERROR! We tried to access a function scoped variable outside the function it's defined in
```

~~~

Take a while to brew on that example. In the end, it's not some mind-blowing concept but there's a whole bunch of terms in there - it'll all help us understand the next mammoth - closures.

### Closures aren't scary

The best way to approach this would be to start with an example - take a look at this piece of code below.

~~~javascript
function makeAdding (first) { // "first" is scoped within the makeAdding function
  return function resulting (second) { // "second" is scoped within the resulting function
    return first + second;
  }

> } // but we've not seen an example of a **function** being returned, thus far - how do we use it?

Reviewer: Let's use quotes instead since md it doesn't work inside a code block
```suggestion
} // but we've not seen an example of a "function" being returned, thus far - how do we use it?
```

const add5 = makeAdding(5);
console.log(add5(2)) // logs 7

> ~~~

Author: gotcha. It's as simple as assigning the arg to a variable, so not concerning.

A lot going on, so let's break it down:

1. The `makeAdding` function takes an argument, `first`, and returns another **function** (which we have named as `add5` while using).
1. When an argument is passed to the returned function (in our case, `add5`), it returns the result of adding up the number passed earlier to the number passed now (`first` to `second`).

Now, while it may sound good at first glance, you can already be raising your eyebrows at the second statement. As we've learned, the `first` variable is scoped within the `makeAdding` function. When we declare and use `add5`, however, we're **outside** the `makeAdding` function. How does the `first` variable still exist, ready to be added when we pass an argument to the `add5` function? This is where we encounter the concept of closures.

> Functions in Javascript form closures. A closure refers to the combination of a function and the **surrounding state** within which a function was declared in. This surrounding state, also called its lexical environment, consists of any local variables that were in scope at the time the closure was made. Here, `add5` is a reference to the `resulting` function, created when the `makeAdding` function is executed, thus it has access to the lexical environment of the `resulting` function, which contains the `first` variable, making it available for use, rather than deleting it for being out of scope.

Reviewer: small typo and I also think `enclosing context` is a better term here
```suggestion
Functions in JavaScript form closures. A closure refers to the combination of a function and the **enclosing context** within which a function was declared in. This enclosing context, also called its lexical environment, consists of any local variables that were in scope at the time the closure was made. Here, `add5` is a reference to the `resulting` function, created when the `makeAdding` function is executed. Therefore, `add5` has access to the lexical environment of the `resulting` function, which contains the `first` variable, making it available for use, rather than deleting it for being out of scope.
```

I also think we can try to elaborate more on this:
> rather than deleting it for being out of scope
"deleting" what? What mechanism does it? What "deletes" it. It's something we can try making a small note of as this kind of behavior is crucial in React as well.

Author: Can you point out what to write instead? Cause I am not sure about how elaborating can be done while keeping it simple to understand. @fcasibu

Reviewer: @Lofty-Brambles I also can't think of a simple explanation to expand on this 😓 but here's what I can think of to rephrase:

```suggestion
Functions in JavaScript form closures. A closure refers to the combination of a function and the **surrounding state** within which a function was declared in. This surrounding state, also called its lexical environment, consists of any local variables that were in scope at the time the closure was made. Here, `add5` is a reference to the `resulting` function, created when the `makeAdding` function is executed. Therefore, instead of being removed from memory due to going out of scope, the `first` variable remains accessible within the lexical environment of the `resulting` function, allowing it to be used by `add5`.
```
What do you think?

Author: LGTM!

While still seen in code, these problems led to the use of a pattern that was similar to constructors but addressed a ton of these problems: Factory Functions.

### Factory functions 🏭

These fancy-sounding functions work very similar to how constructors did, but with one key difference - they levy the power of closures. Instead of using the `new` keyword to create an object, factory functions simply set up and return the new object when you call the function. They do not use the prototype, which incurs a performance penalty - but as a general rule, this penalty isn’t significant unless you’re creating thousands of objects. Let's take a basic example to compare them to constructor functions.

~~~javascript
const User = function (name) {
  this.name = name;
  this.discordName = "@" + name;
} // hey, this is a constructor - then this can be refactored into a factory!

function createUser (name) {
  const discordName = "@" + name;
  return { name, discordName };
} // and that's very similar, except since it's just a function, we don't need a new keyword

> ~~~

Author: Aight, got that. Should be an easy fix, so I'll do it now.

> ~~~

Author: (This preview)[https://www.theodinproject.com/lessons/preview?uuid=b43a6854-1fe9-4419-b967-8c2b8ebf0bad] should be a good demo

In this case, we did not need control of `reputation` itself. To avoid foot guns, like accidentally setting the reputation to `-18000`, we simply expose the necessary details in the form of `getReputation` and `giveReputation`.

### Prototypal inheritance with factories

In the lesson with constructors, we looked deeply into the concept of prototype and inheritance, and how to give our objects access to the properties of another. With factory functions too, there are easy ways to do that. Take another hypothetical scenario into consideration. We need to extend the `User` factory into a `Player` factory that needs to control some more metrics - there are some ways to do that:

~~~javascript
function createPlayer (name, level) {

>   const { discordName, getReputation } = createUser(name);

Reviewer: Since destructuring isn't mentioned anywhere in the previous lessons, I think we should also add the same comment here for clarity.

Author: Would a note on this, separately, be a better idea? Like the one with object shorthand.

Reviewer 2: I would probably just explain it along with the object shorthand stuff. I feel like the concepts play off of each other.

Take the calculator example into consideration. It's very easy to imagine a scenario where you can accidentally create multiple functions with the name `add`. What does `add` do - does it simply add two numbers? Strings? Does it take its input directly from the DOM and display the result? What would you name the functions that do these things? Instead, we can easily encapsulate them inside a module called `calculator` which generates an object with that name, allowing us to explicitly call `calculator.add(a, b)` or `calculator.sub(a, b)`.

### Assignment

<div class="lesson-content__panel" markdown="1">

1. WesBos has a beautiful and in-depth section on scopes and closures. Please check out these sections under "Module 3 - The Tricky Bits":
  - [The article on scope](https://wesbos.com/javascript/03-the-tricky-bits/scope)
  - [The article on closures](https://wesbos.com/javascript/03-the-tricky-bits/closures)
1. [This article](https://tsherif.wordpress.com/2013/08/04/constructors-are-bad-for-javascript/) goes into depth, discussing the problems with constructors as well as suggesting factories as a solution.

> 1. [This article](https://dev.to/tomekbuszewski/module-pattern-in-javascript-56jm) provides more details about module pattern with examples.

Reviewer 2: Could you tweak the link text on these to be a bit more a11y friendly?

Author: Done!
